integration/image/delta_test.go

```diff
@@ -0,0 +1,214 @@
+	type testCase struct {
+		desc           string
+		base           string
+		target         string
+		delta          string
+		expectedImages []string
+	}
+
+	for _, c := range []testCase{
+		{
+			desc:           "busybox images",
+			base:           "busybox:1.24",
+			target:         "busybox:1.29",
+			delta:          fmt.Sprintf("%s/delta-test:1.24-1.29", registryURI),
+			expectedImages: []string{"busybox:1.24", fmt.Sprintf("%s/delta-test:1.24-1.29", registryURI)},
+		},
+	} {
+		t.Run(c.desc, func(t *testing.T) {
+			c := c
+
+			var (
+				err error
+				rc  io.ReadCloser
+			)
+
+			d := daemon.New(t, "", "balena-engine-daemon", daemon.Config{})
+			client, err := d.NewClient()
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			var args = []string{}
+
+			d.Start(t, args...)
+			ctx := context.Background()
+
+			reg, err := registry.NewV2(false, "htpasswd", "", registryURI)
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer reg.Close()
+
+			t.Log("Pulling delta base")
+			rc, err = client.ImagePull(ctx,
+				c.base,
+				types.ImagePullOptions{})
+			if err != nil {
+				t.Fatal(err)
+			}
+			io.Copy(ioutil.Discard, rc)
+			rc.Close()
+
+			t.Log("Pulling delta target")
+			rc, err = client.ImagePull(ctx,
+				c.target,
+				types.ImagePullOptions{})
+			if err != nil {
+				t.Fatal(err)
+			}
+			io.Copy(ioutil.Discard, rc)
+			rc.Close()
+
+			t.Log("Creating delta")
+			var deltaID string
+			rc, err = client.ImageDelta(ctx,
+				c.base,
+				c.target)
+			if err != nil {
+				t.Fatal(err)
+			}
+			{
+				re := regexp.MustCompile("sha256:([a-fA-F0-9]{64})$")
+				sc := bufio.NewScanner(rc)
+				for sc.Scan() {
+					var v map[string]interface{}
+					if err := json.Unmarshal(sc.Bytes(), &v); err != nil {
+						t.Fatal(err)
+					}
+					if _, ok := v["errorDetail"]; ok {
+						t.Fatal(errors.New(v["errorDetail"].(string)))
+					}
+					_, ok := v["progressDetail"]
+					if ok {
+						continue
+					}
+					s, ok := v["status"]
+					if !ok {
+						continue
+					}
+					status, ok := s.(string)
+					if !ok {
+						continue
+					}
+					r := re.FindStringSubmatch(status)
+					if len(r) < 1 {
+						continue
+					}
+					deltaID = r[0]
+				}
+				if err := sc.Err(); err != nil {
+					t.Fatal(err)
+				}
+				if deltaID == "" {
+					t.Fatal("Unable to parse delta image id from progress output")
+				}
+			}
+			rc.Close()
+
+			err = client.ImageTag(ctx, deltaID, c.delta)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			t.Log("Pushing delta to local registry")
+			encodedAuth, err := command.EncodeAuthToBase64(types.AuthConfig{
+				Username: reg.Username(),
+				Password: reg.Password(),
+				Auth:     "htpasswd",
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+			rc, err = client.ImagePush(ctx, c.delta, types.ImagePushOptions{
+				RegistryAuth: encodedAuth,
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+			io.Copy(ioutil.Discard, rc)
+			rc.Close()
+
+			t.Log("Stopping daemon")
+			d.Stop(t)
```

Comment:
If there is an error between the New/Start and this Stop call, like one of the t.Fatal(err) calls above, what happens to the daemon? Is it automatically (and quickly/soon) garbage collected, or does it stay around using resources? (The integration tests take 2 hours on my laptop, resources could be held for such a long time.)

By the way, I assume that we want for it to be possible "to run multiple runs" of the integration tests in parallel in the same build server. Meaning, for example, two independent pull requests triggering the tests more or less at the same time. Does `TestDelta` (including the docker daemon) create any "singleton" resources like, say, a lock file in `/tmp/delta_lock_file` or a database table that would be common across tests runs and interfere with each other? In which case we would want to namespace the resources, like appending the build number to their name: `/tmp/delta_lock_file_build_389221`.
(I certainly wouldn't insist on this if you could show that several other existing tests already create singleton resources that prevent running integration tests in parallel, so that TestDelta does no worse...)

Reply:
> If there is an error between the New/Start and this Stop call, like one of the t.Fatal(err) calls above, what happens to the daemon?

I will try to just use configReload which might be fine in our case here, since I only restart the daemon to get a new root store.

> Meaning, for example, two independent pull requests triggering the tests more or less at the same time. Does TestDelta (including the docker daemon) create any "singleton" resources [...]

That should be fine since the test daemon is started inside the test container. For parallel test  runs the daemon creates independent daemon dirs (this works for the container restart test already https://github.com/balena-os/balena-engine/blob/17.12-resin/integration/container/restart_test.go)

Reply:
`daemon.ConfigReload()` doesn't seem to create new root directories so it's not an option

To solve the daemon gc I defer the first `Stop()` instead.
<!-- https://play.golang.org/p/Jotuse-G9Zy -->
In case the test fails somewhere in between that should cleanup whichever instance of it is running at that point.